### ISSUE TITLE
Scale-up goes all the way to 100% immediately after upgrade from 4.5.0 to 5.3.1

### ISSUE THREAD

**Issue by DanielHeath** [buildkite/elastic-ci-stack-for-aws#847]
I have my ASG set to max capacity 2, minimum 0.

Previously, when a build came in, the desired capacity for the stack would be set to 1; now, it increases to 1 and then again to 2 before the first job starts.

ASG Timeline:
![image](https://user-images.githubusercontent.com/47430/118428673-5435b680-b713-11eb-91ba-dcc5e680147e.png)
It seems the stack only waited 10s after the 0 -> 1 increase to go 1 -> 2.

Build timeline:
```
2021-05-17 03:12:16 UTC Running global environment hook
2021-05-17 03:12:31 UTC | 2021-05-17 03:12:31 INFO   Successfully uploaded and parsed pipeline config
```

According to the buildkite UI, the job only waited 1m19s.

Non-sensitive params as follows:
```
AgentsPerInstance 15  -
ArtifactsBucket - -
AssociatePublicIpAddress  true  -
AuthorizedUsersUrl  - -
AvailabilityZones - -
BuildkiteAdditionalSudoPermissions  - -
BuildkiteAgentExperiments ansi-timestamps,git-mirrors,resolve-commit-after-checkout,normalised-upload-paths -
BuildkiteAgentRelease stable  -
BuildkiteAgentTags  - -
BuildkiteAgentTimestampLines  true  -
BuildkiteAgentTokenParameterStoreKMSKey - -
BuildkiteAgentTokenParameterStorePath - -
BuildkiteQueue  default -
BuildkiteTerminateInstanceAfterJob  false -
BuildkiteWindowsAdministrator true  -
CostAllocationTagName aws:createdBy -
CostAllocationTagValue  buildkite-elastic-ci-stack-for-aws  -
ECRAccessPolicy full  -
EnableAgentGitMirrorsExperiment true  -
EnableCostAllocationTags  false -
EnableDockerExperimental  false -
EnableDockerLoginPlugin true  -
EnableDockerUserNamespaceRemap  false -
EnableECRPlugin true  -
EnableSecretsPlugin true  -
IMDSv2Tokens  optional  -
ImageId - -
ImageIdParameter  - -
InstanceCreationTimeout PT5M  -
InstanceOperatingSystem linux -
InstanceRoleName  - -
InstanceRolePermissionsBoundaryARN  - -
InstanceType  c5.9xlarge  -
ManagedPolicyARN  - -
MaxSize 2 -
MinSize 0 -
OnDemandPercentage  0 -
RootVolumeName  /dev/xvda -
RootVolumeSize  80  -
RootVolumeType  gp3 -
ScaleInIdlePeriod 120 -
ScaleOutFactor  0.4 -
ScaleOutForWaitingJobs  false -
SecurityGroupId - -
SpotPrice 1.30  -
Subnets - -
VpcId -
```

**Comment by yob** [comment]
Hi Daniel,

Are you able to check if 5.2.0 has the same issue?

I suspect it won't, and that what you're seeing is the anti-stall behavior
we added in 5.3.0. Possibly a bug in said behaviour.

**Comment by DanielHeath** [comment]
I'm not likely to look too soon, as it's not costing a fortune right now.

**Comment by yob** [comment]
no worries.

@dbaggerman - do you think this could be the changes in https://github.com/buildkite/buildkite-agent-scaler/pull/40 ?

**Comment by dbaggerman** [comment]
It's certainly possible. Now that you mention it, I had meant to look into how long it takes between changing the `DesiredCount` number and actually seeing an instance in a `pending` state in the list.

If the new instance isn't in the ASG instance list by the time the next time round the polling loop, then it won't count the pending agents to the anticipated total. That might cause it to jump in and increment the desired count again in an attempt to compensate.

**Comment by theonlysinjin** [comment]
So I have experienced similar... I've set the ASG min to 0 and max to 3.
I'll post the lambda logs when I next dig, but, something like the following is happening
- Job Gets Queued
- Instance starts due to the agent-scaler
- about 10s later a second instance starts, also due to the scaler must think nothing has happened? But in the same run?
- Then the next time the function runs, it scales _again_, pushing this up to 3 instances.

This is happening across all our stacks. But it's like the scaler can't detect that it's just scaled an ASG.

**Comment by jessebye** [comment]
We're impacted by this as well. We have a min of 1, max of 50. When a new job comes in, it doesn't schedule on the 1 available agent, but instead waits until the ASG scales to 50. Then the job runs and the ASG scales back down to 1.

**Comment by yob** [comment]
Thanks for your patience folks - we've released 5.3.2 which we believe fixes this issue :heart:

**Comment by timatooth** [comment]
🥳
